File: LeetcodeSwift/LeetcodeSwift/algorithms/q0437.swift
//
//  q0437.swift
//  LeetcodeSwift
//
//  Created by NowOrNever on 18/02/2020.
//  Copyright © 2020 DL. All rights reserved.
//

import Foundation
//437. Path Sum III
//Easy
//
//2693
//
//197
//
//Add to List
//
//Share
//You are given a binary tree in which each node contains an integer value.
//
//Find the number of paths that sum to a given value.
//
//The path does not need to start or end at the root or a leaf, but it must go downwards (traveling only from parent nodes to child nodes).
//
//The tree has no more than 1,000 nodes and the values are in the range -1,000,000 to 1,000,000.
//
//Example:
//
//root = [10,5,-3,3,2,null,11,3,-2,null,1], sum = 8
//
//      10
//     /  \
//    5   -3
//   / \    \
//  3   2   11
// / \   \
//3  -2   1
//
//Return 3. The paths that sum to 8 are:
//
//1.  5 -> 3
//2.  5 -> 2 -> 1
//3. -3 -> 11
class q0437Solution {
    func pathSum_helper(node: TreeNode?, sum:Int) -> Int{
        guard let node = node else { return 0 }
        return (node.val == sum ? 1 : 0) + pathSum_helper(node: node.left, sum: sum - node.val) + pathSum_helper(node: node.right, sum: sum - node.val)
    }
    func pathSum(_ root: TreeNode?, _ sum: Int) -> Int {
        guard let root = root else { return 0 }
        return pathSum_helper(node: root, sum: sum) + pathSum(root.left, sum) + pathSum(root.right, sum)
    }
}

File: LeetcodePython3/q0437.py
from typing import Dict


class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right

class Solution:
    result = 0
    def pathSum_helper(self, root: TreeNode, sumDic: Dict, sum:int):
        if not root:
            return
        if root.val in sumDic:
            self.result += sumDic[root.val]
        if sum == root.val:
            self.result += 1

        newDic = {}
        for key in sumDic.keys():
            newDic[key - root.val] = sumDic[key]
        if (sum - root.val) not in newDic:
            newDic[sum - root.val] = 1
        else:
            newDic[sum - root.val] += 1

        self.pathSum_helper(root.left, newDic, sum)
        self.pathSum_helper(root.right, newDic, sum)

    def pathSum(self, root: TreeNode, sum: int) -> int:
        self.result = 0
        sumDic = {}
        self.pathSum_helper(root, sumDic, sum)
        return self.result



